components/automate-gateway/gateway/server.go

```diff
@@ -325,42 +125,314 @@ func (s *Server) NewGRPCServer() (*grpc.Server, error) {
+	if err != nil {
+		return errors.Wrap(err, "create authz client")
+	}
+	authzClientV2, err := s.clientsFactory.AuthorizationV2Client()
+	if err != nil {
+		return errors.Wrap(err, "create authz_v2 client")
+	}
+
+	s.authorizer = authorizer.NewAuthorizer(
+		authv1.AuthorizationHandler(authzClientV1),
+		authv2.AuthorizationHandler(authzClientV2),
+	)
+
+	return nil
+}
+
+// startNullBackendServer starts the unimplemented backend server
+func (s *Server) startNullBackendServer() error {
+	var err error
+
+	s.logger.Info("starting null backend server")
+
+	nullBackendListener, err := net.Listen("unix", s.Config.GrpcClients.NullBackendSock)
+	if err != nil {
+		return errors.Wrapf(err, "listen on %s", s.Config.GrpcClients.NullBackendSock)
+	}
+
+	s.nullBackendServer = nullbackend.NewServer()
+
+	// Start all servers in goroutines and start watching for user signals. If
+	// an error is returned from a server or an exit signal is received, try to
+	// gracefully stop servers and exit.
+	go func() {
+		err := s.nullBackendServer.Serve(nullBackendListener)
+		if err != nil {
+			s.errC <- errors.Wrap(err, "serve null backend")
+		}
+		return
+	}()
+
+	return nil
+}
+
+func (s *Server) stopNullBackendServer() {
+	s.logger.Info("stopping null backend server")
+
+	s.nullBackendServer.GracefulStop()
+	_ = os.Remove(s.Config.GrpcClients.NullBackendSock)
+
+	return
+}
+
+func (s *Server) startSignalHandler() error {
+	s.logger.Info("starting signal handlers")
+
+	for {
+		select {
+		case err := <-s.errC:
+			switch errors.Cause(err) {
+			// ErrServerClosed is returned if an HTTP server is shutdown, which
+			// can only happen if it's triggered by a shutdown signal or a
+			// reconfigure signal, in which case the server shutting down is
+			// intended. The shutdown error is only returned nothing goes wrong
+			// when shutting it down, therefore we'll log it and move on.
+			case http.ErrServerClosed:
+				s.logger.Debug("HTTP server was recently shutdown")
+			default:
+				s.logger.WithError(err).Error("exiting")
+				err1 := s.stop()
+				if err1 != nil {
+					return errors.Wrap(err, err1.Error())
+				}
+				return err
+			}
+		case sig := <-s.sigC:
+			switch sig {
+			case syscall.SIGHUP:
+				s.logger.WithField("signal", sig).Info("handling received signal")
+				// Reconfigure
+				err := s.reconfigure()
+				if err != nil {
+					err2 := s.stop()
+					if err2 != nil {
+						err = errors.Wrap(err, err2.Error())
+					}
+
+					return err
+				}
+			case syscall.SIGUSR1, syscall.SIGTERM, syscall.SIGINT:
+				s.logger.WithField("signal", sig).Info("handling received signal")
+				return s.stop()
+			default:
+				s.logger.WithField("signal", sig).Warn("unable to handle received signal")
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *Server) stop() error {
+	s.logger.Info("stopping automate-gateway")
+
+	err := s.stopHTTPServer()
+
+	s.stopGRPCServer()
+	s.stopNullBackendServer()
+
+	return err
+}
+
+func (s *Server) reconfigure() error {
```

Comment:
I might be missing something obvious, but why doesn't this end up leaking memory. Does everything we use end up getting garbage collected if its no longer accessible? I feel like the notifications client was one of the things that ended up leaking a bunch of memory when it was accidentally getting recreated a lot